### PR TITLE
Initial commit for assignment 1.2

### LINKED ISSUE #1: Assignment 1.2
In this assignment , the very basics of git are put under scrutiny

### PR DESCRIPTION
**Pull request by YashwanthASN-TA**
Closes #1

### CODE REVIEW

File: nonstandardcode.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import tarfile
+from six.moves import urllib
+
+
+DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml/master/"
+HOUSING_PATH = os.path.join("datasets", "housing")
+HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz"
+
+def fetch_housing_data(housing_url=HOUSING_URL, housing_path=HOUSING_PATH):
+    os.makedirs(housing_path, exist_ok=True)
```
Comment (yashwanttiger7225) [inline nonstandardcode.py]: please add docstring

File: nonstandardcode.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    tgz_path = os.path.join(housing_path, "housing.tgz")
+    urllib.request.urlretrieve(housing_url, tgz_path)
+    housing_tgz = tarfile.open(tgz_path)
+    housing_tgz.extractall(path=housing_path)
+    housing_tgz.close()
+
+import pandas as pd
```
Comment (yashwanttiger7225) [inline nonstandardcode.py]: place import to the top of the file

File: nonstandardcode.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+param_grid = [
+    # try 12 (3×4) combinations of hyperparameters
+    {'n_estimators': [3, 10, 30], 'max_features': [2, 4, 6, 8]},
+    # then try 6 (2×3) combinations with bootstrap set as False
+    {'bootstrap': [False], 'n_estimators': [3, 10], 'max_features': [2, 3, 4]},
+  ]
+
+forest_reg = RandomForestRegressor(random_state=42)
+# train across 5 folds, that's a total of (12+6)*5=90 rounds of training
```
Comment (yashwanttiger7225) [inline nonstandardcode.py]: add comments like this after each useful operations

File: nonstandardcode.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    # try 12 (3×4) combinations of hyperparameters
+    {'n_estimators': [3, 10, 30], 'max_features': [2, 4, 6, 8]},
+    # then try 6 (2×3) combinations with bootstrap set as False
+    {'bootstrap': [False], 'n_estimators': [3, 10], 'max_features': [2, 3, 4]},
+  ]
+
+forest_reg = RandomForestRegressor(random_state=42)
+# train across 5 folds, that's a total of (12+6)*5=90 rounds of training 
+grid_search = GridSearchCV(forest_reg, param_grid, cv=5,
+                           scoring='neg_mean_squared_error', return_train_score=True)
```
Comment (yashwanttiger7225) [inline nonstandardcode.py:246]: move this values into config file if someone wants to change can be change from config file

**Review by yashwanttiger7225 — CHANGES_REQUESTED** [review]

File: nonstandardcode.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+
+# Define the root URL for downloading the dataset and set up paths for storing it
+DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml/master/"
+HOUSING_PATH = os.path.join("datasets", "housing")
+HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz"
+
+def fetch_housing_data(housing_url=HOUSING_URL, housing_path=HOUSING_PATH):
+    """
+    Fetch the housing data from the provided URL and extract it to the specified path.
```
Comment (yashwanttiger7225) [inline nonstandardcode.py:29]: please define args and return

File: nonstandardcode.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    os.makedirs(housing_path, exist_ok=True)  # Create the directory if it doesn't exist
+    tgz_path = os.path.join(housing_path, "housing.tgz")  # Path for the .tgz file
+    urllib.request.urlretrieve(housing_url, tgz_path)  # Download the .tgz file
+    housing_tgz = tarfile.open(tgz_path)  # Open the .tgz file
+    housing_tgz.extractall(path=housing_path)  # Extract all contents of the file
+    housing_tgz.close()  # Close the .tgz file
+
+def load_housing_data(housing_path=HOUSING_PATH):
+    """
+    Load the housing data as a pandas DataFrame.
```
Comment (yashwanttiger7225) [inline nonstandardcode.py]: add here as well

File: nonstandardcode.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+# StratifiedShuffleSplit for splitting the data while preserving income category proportions
+split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
+for train_index, test_index in split.split(housing, housing["income_cat"]):
+    strat_train_set = housing.loc[train_index]  # Training data after stratification
+    strat_test_set = housing.loc[test_index]  # Test data after stratification
+
+def income_cat_proportions(data):
+    """
+    Calculate income category proportions.
+    """
```
Comment (yashwanttiger7225) [inline nonstandardcode.py:107]: here

File: nonstandardcode.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+def fetch_housing_data(housing_url=HOUSING_URL, housing_path=HOUSING_PATH):
+    """
+    Fetch the housing data from the provided URL and extract it to the specified path.
+
+    This function downloads the housing dataset from the given URL, and extracts it 
+    into the specified directory (housing_path). If the directory doesn't exist, it 
+    creates the directory.
+
+    Parameters:
```
Comment (yashwanttiger7225) [inline nonstandardcode.py]: use Args insted of parameters..

File: env.yml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+      - pandas==2.2.3
+      - python-dateutil==2.9.0.post0
+      - pytz==2024.2
+      - scikit-learn==1.5.2
+      - scipy==1.14.1
+      - six==1.16.0
+      - threadpoolctl==3.5.0
+      - tzdata==2024.2
+
+
```
Comment (yashwanttiger7225) [inline env.yml:17]: why this black lines?

**Review by yashwanttiger7225 — CHANGES_REQUESTED** [review]

File: env.yml (hunk 9 above)
Comment (YashwanthASN-TA) [reply env.yml:17]: Removed the blank lines

File: nonstandardcode.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+    housing_tgz = tarfile.open(tgz_path)  # Open the .tgz file
+    housing_tgz.extractall(path=housing_path)  # Extract all contents of the file
+    housing_tgz.close()  # Close the .tgz file
+
+def load_housing_data(housing_path=HOUSING_PATH):
+    """
+    Calculate the proportion of each income category in the provided DataFrame.
+
+    This function takes a DataFrame as input and computes the relative proportions
+    of different income categories defined in the 'income_cat' column. The proportions
```
Comment (yashwanttiger7225) [inline nonstandardcode.py]: can you do properly this doc string doesnt match with functionality of function where are you using housing path? and where is incom_cat. this is really a wrong way to do that.

**Review by yashwanttiger7225 — CHANGES_REQUESTED** [review]

**Review by yashwanttiger7225 — APPROVED** [review]

**Pull request merged**